src/chronify/time_series_mapper_representative.py

```diff
@@ -0,0 +1,140 @@
+        """Check that from_schema can produce to_schema."""
+        self._check_table_column_producibility()
+        self._check_schema_measurement_type_consistency()
+
+    def _check_table_column_producibility(self) -> None:
+        """Check columns in destination table can be produced by source table."""
+        available_cols = self._from_schema.list_columns() + [
+            self._to_schema.time_config.time_column
+        ]
+        final_cols = self._to_schema.list_columns()
+        if diff := set(final_cols) - set(available_cols):
+            msg = f"Source table {self._from_schema.name} cannot produce the columns: {diff}"
+            raise ConflictingInputsError(msg)
+
+    def _check_schema_measurement_type_consistency(self) -> None:
+        """Check that measurement_type is the same between schema."""
+        from_mt = self._from_schema.time_config.measurement_type
+        to_mt = self._to_schema.time_config.measurement_type
+        if from_mt != to_mt:
+            msg = f"Inconsistent measurement_types {from_mt=} vs. {to_mt=}"
+            raise ConflictingInputsError(msg)
+
+    def _check_source_table_has_time_zone(self) -> None:
+        """Check source table has time_zone column."""
+        if "time_zone" not in self._from_schema.list_columns():
+            msg = f"time_zone is required for tz-aware representative time mapping and it is missing from source table: {self._from_schema.name}"
+            raise MissingParameter(msg)
+
+    def map_time(self) -> None:
+        """Convert time columns with from_schema to to_schema configuration."""
+
+        is_tz_naive = self._to_schema.time_config.is_time_zone_naive()
+        self.check_schema_consistency()
+        if not is_tz_naive:
+            self._check_source_table_has_time_zone()
+        # TODO: add interval type handling (note annual has no interval type)
+
+        map_table_name = "map_table"
+        dfm = self._create_mapping(is_tz_naive)
+        if map_table_name in self._metadata.tables:
+            msg = (
+                f"table {map_table_name} already exists, delete it or use a different table name."
+            )
+            raise TableAlreadyExists(msg)
+        try:
+            with self._engine.connect() as conn:
+                write_database(dfm, conn, map_table_name, self._to_schema.time_config)
+                conn.commit()
+            self._metadata.reflect(self._engine, views=True)
+            self._apply_mapping(map_table_name)
+
+            mapped_table = Table(self._to_schema.name, self._metadata)
+            with self._engine.connect() as conn:
+                try:
+                    check_timestamps(conn, mapped_table, self._to_schema)
+                except Exception as e:
+                    print(e)
+                    print(f"check_timestamps failed, dropping mapped table {self._to_schema.name}")
+                    conn.execute(text(f"DROP TABLE {self._to_schema.name}"))
```

Comment:
1. You need to call `raise` after this line. Otherwise, the user will think it worked.
2. Change print statements to `logger.exception("...")`. It will log the backtrace.
3. Change line 81 to `except Exception`.
4. Can `self._apply_mapping` run inside the context manager on line 71? If so, you can call `conn.rollback` instead of deleting the table. I'm not sure that the `create_table` will work under the same connection, though. Ideally, there is one connection, one commit, and one optional rollback.

Reply:
I tested the changes below. It works. Something that you probably haven't noticed is that when you call `DROP TABLE {map_table_name}`, it does not delete the table from `self._metadata`. I corrected that below. For some reason, that causes the next `map_time` call to fail. I'll look at that next.
```python
        map_table = Table(map_table_name, self._metadata)
        try:
            with self._engine.connect() as conn:
                write_database(dfm, conn, map_table_name, self._to_schema.time_config)
                self._metadata.reflect(self._engine, views=True)
                self._apply_mapping(map_table_name)
                mapped_table = Table(self._to_schema.name, self._metadata)
                try:
                    check_timestamps(conn, mapped_table, self._to_schema)
                except Exception:
                    logger.exception(
                        "check_timestamps failed on mapped table {}. Drop it.", self._to_schema.name
                    )
                    conn.rollback()
                    raise
                conn.commit()
        finally:
            self._metadata.drop_all(self._engine, tables=[map_table])
            self._metadata.remove(map_table)

```

Reply:
Thanks